commit aa333c020dc3f145dccec85535b0e80985eff67b
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2019-03-07T00:22:38+00:00 (7 months ago)
    
    MANTA-4144 Add metricPorts mdata variable to pgstatsmon
    Reviewed by: Kody Kantor <kody.kantor@joyent.com>
    Approved by: Kody Kantor <kody.kantor@joyent.com>

diff --git a/Makefile b/Makefile
index a9f3a92..7aff5bb 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2018, Joyent, Inc. All rights reserved.
+# Copyright (c) 2019, Joyent, Inc. All rights reserved.
 #
 # Makefile: pgstatsmon - Postgres monitoring system
 #
@@ -81,8 +81,11 @@ NODE_BITS_DIR	= $(PREFIX)/node
 SAPI_MANIFESTS		= pgstatsmon
 SAPI_MANIFESTS_DIRS	= $(SAPI_MANIFESTS:%=$(PREFIX)/sapi_manifests/%)
 
-SMF_MANIFEST		= pgstatsmon
+SMF_MANIFESTS		= pgstatsmon metric-ports-updater
+SMF_METHODS		= metric-ports-updater
+
 SMF_MANIFEST_DIR	= $(PREFIX)/smf/manifests
+SMF_METHOD_DIR		= $(PREFIX)/smf/methods
 
 DTRACE_SCRIPTS		= backendstat.d querystat.d walstat.d
 DTRACE_SCRIPTS_DIR	= $(PREFIX)/bin/dtrace
@@ -99,7 +102,8 @@ INSTALL_FILES	= $(addprefix $(PROTO), \
 		  $(BOOT_SCRIPTS:%=$(BOOT_SCRIPTS_DIR)/%) \
 		  $(SAPI_MANIFESTS_DIRS:%=%/template) \
 		  $(SAPI_MANIFESTS_DIRS:%=%/manifest.json) \
-		  $(SMF_MANIFEST:%=$(SMF_MANIFEST_DIR)/%.xml) \
+		  $(SMF_MANIFESTS:%=$(SMF_MANIFEST_DIR)/%.xml) \
+		  $(SMF_METHODS:%=$(SMF_METHOD_DIR)/%.sh) \
 		  )
 
 INSTALL_DIRS	= $(addprefix $(PROTO), \
@@ -113,6 +117,7 @@ INSTALL_DIRS	= $(addprefix $(PROTO), \
 		  $(NODE_BITS_DIR)/lib \
 		  $(BOOT_SCRIPTS_DIR) \
 		  $(SMF_MANIFEST_DIR) \
+		  $(SMF_METHOD_DIR) \
 		  $(SAPI_MANIFESTS_DIRS) \
 		  )
 
@@ -217,6 +222,10 @@ $(PROTO)$(PREFIX)/sapi_manifests/%: sapi_manifests/% | $(INSTALL_DIRS)
 $(PROTO)$(PREFIX)/smf/manifests/%: smf/manifests/% | $(INSTALL_DIRS)
 	$(INSTALL_FILE)
 
+# install SMF methods
+$(PROTO)$(PREFIX)/smf/methods/%: smf/methods/% | $(INSTALL_DIRS)
+	$(INSTALL_FILE)
+
 include ./tools/mk/Makefile.targ
 include ./tools/mk/Makefile.node_modules.targ
 include ./tools/mk/Makefile.node_prebuilt.targ
diff --git a/boot/firstboot.sh b/boot/firstboot.sh
index de42ae7..270c057 100644
--- a/boot/firstboot.sh
+++ b/boot/firstboot.sh
@@ -6,20 +6,21 @@
 #
 
 #
-# Copyright (c) 2018, Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
-printf '==> firstboot @ %s\n' "$(date -u +%FT%TZ)"
+printf "==> firstboot @ %s\n" "$(date -u +%FT%TZ)"
 
 set -o xtrace
 
-NAME=pgstatsmon
+NAME="pgstatsmon"
 
 #
 # Runs on first boot of a newly reprovisioned "pgstatsmon" zone.
 #
 
 SVC_ROOT="/opt/smartdc/$NAME"
+METRICPORTS_SVC="metric-ports-updater"
 
 #
 # Build PATH from this list of directories.  This PATH will be used both in the
@@ -40,7 +41,7 @@ paths=(
 PATH=
 for (( i = 0; i < ${#paths[@]}; i++ )); do
 	if (( i > 0 )); then
-		PATH+=':'
+		PATH+=":"
 	fi
 	PATH+="${paths[$i]}"
 done
@@ -61,8 +62,8 @@ manta_common_setup "$NAME"
 # Replace the contents of PATH from the default root user .bashrc with one
 # more appropriate for this particular zone.
 #
-if ! /usr/bin/ed -s '/root/.bashrc'; then
-	fatal 'could not modify .bashrc'
+if ! /usr/bin/ed -s "/root/.bashrc"; then
+	fatal "could not modify .bashrc"
 fi <<EDSCRIPT
 /export PATH/d
 a
@@ -71,12 +72,19 @@ export PATH="$PATH"
 w
 EDSCRIPT
 
+#
+# Import the metric-ports-updater SMF service.
+#
+if ! svccfg import "/opt/smartdc/$NAME/smf/manifests/$METRICPORTS_SVC.xml"; then
+	fatal "could not import $METRICPORTS_SVC SMF service"
+fi
+
 #
 # Import the pgstatsmon SMF service.  The manifest file creates the service
 # enabled by default.
 #
 if ! svccfg import "/opt/smartdc/$NAME/smf/manifests/$NAME.xml"; then
-	fatal 'could not import SMF service'
+	fatal "could not import $NAME SMF service"
 fi
 
 manta_common_setup_end
diff --git a/smf/manifests/metric-ports-updater.xml b/smf/manifests/metric-ports-updater.xml
new file mode 100644
index 0000000..2d712d8
--- /dev/null
+++ b/smf/manifests/metric-ports-updater.xml
@@ -0,0 +1,54 @@
+<?xml version='1.0'?>
+<!DOCTYPE service_bundle SYSTEM '/usr/share/lib/xml/dtd/service_bundle.dtd.1'>
+<!--
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+-->
+
+<!--
+    Copyright (c) 2019, Joyent, Inc.
+-->
+
+<service_bundle type='manifest' name='manta-application-metric-ports-updater'>
+    <service name='manta/application/metric-ports-updater'
+             type='service'
+             version='1'>
+        <create_default_instance enabled='true'/>
+        <single_instance/>
+
+        <dependency name='pgstatsmon'
+                    grouping='optional_all'
+                    restart_on='refresh'
+                    type='service'>
+            <service_fmri value='svc:/manta/application/pgstatsmon'/>
+        </dependency>
+
+        <exec_method type='method'
+                     name='start'
+                     exec='/bin/bash /opt/smartdc/pgstatsmon/smf/methods/metric-ports-updater.sh'
+                     timeout_seconds='30'>
+            <method_context working_directory='/opt/smartdc/pgstatsmon'>
+                <method_credential user='root' group='root'/>
+            </method_context>
+        </exec_method>
+
+        <exec_method name='stop'
+                     type='method'
+                     exec=':true'
+                     timeout_seconds='300'/>
+
+        <property_group name='startd'
+                        type='framework'>
+            <propval name='duration'
+                     type='astring'
+                     value='transient' />
+        </property_group>
+
+        <template>
+            <common_name>
+                <loctext xml:lang='C'>Set metricPorts mdata variable</loctext>
+            </common_name>
+        </template>
+    </service>
+</service_bundle>
diff --git a/smf/methods/metric-ports-updater.sh b/smf/methods/metric-ports-updater.sh
new file mode 100644
index 0000000..a362680
--- /dev/null
+++ b/smf/methods/metric-ports-updater.sh
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright (c) 2019, Joyent, Inc.
+#
+
+#
+# This script updates the zone's metricPorts mdata variable with the current
+# port value from pgstatsmon's config file, with the intent of keeping the mdata
+# value in sync with the port currently being used by pgstatsmon. To this end,
+# this script's SMF manifest is configured to run the script every time the
+# pgstatsmon service is refreshed or restarted.
+#
+
+SVC_NAME=pgstatsmon
+SVC_ROOT="/opt/smartdc/$SVC_NAME"
+SAPI_CONFIG="$SVC_ROOT/etc/config.json"
+
+set -o errexit
+set -o pipefail
+set -o xtrace
+
+if ! source "$SVC_ROOT/scripts/util.sh"; then
+    exit 1
+fi
+
+# Get the metricPort from the sapi config file to allow scraping by cmon-agent.
+#
+# Note: If the config file doesn't exist, the script will fail - but so will
+# pgstatsmon itself, because it depends on the existence of the config file to
+# run.
+if [[ ! -f "$SAPI_CONFIG" ]]; then
+    fatal "SAPI config not found"
+fi
+mdata-put metricPorts $(< "$SAPI_CONFIG" json "target.port")
